File: lens/lens.cpp
// Inspired by https://www.shadertoy.com/view/3sfGzB
#include <stdio.h>
#include <stdlib.h>
#include <conio.h>
#include <math.h>
#include <iostream.h>
#include "vga.h"
#include "keyb.h"
#include "vidbuf.h"
#include "cat.h"

void randomLines() {
  for (int i = 0; i < 200; i++) {
    line(rand()%320, rand()%200,
         rand()%320, rand()%200, 1 + rand()%255);
  }
}

void drawCat() {
  outp(0x03c8, 0);
  for (int i = 0; i < 256; i++) {
    outp(0x03c9, picMap[i*4 + 0]/4);
    outp(0x03c9, picMap[i*4 + 1]/4);
    outp(0x03c9, picMap[i*4 + 2]/4);
  }

  for (unsigned int y = 0; y < 200; y++) {
    for (int x = 0; x < 320; x++) {
      buffer[x + y*320] = picMap[256*4 + x + y*320];
    }
  }
}

int main (void)
{
  if (!buffer) {
    printf("Video buffer is not initialized");
    return 1;
  }

  vgaInit();
  fillBuffer(0);
  drawCat();

  float an1 = 0;
  float an2 = 0;
  const int radx = 40;
  const int rady = 40;
  do {
    flushBuffer(160 + cos(an1)*radx,
                100 + sin(an2)*rady,
                60, 60, -8, -8);
    an1 += 0.07;
    an2 += 0.13;
    // Exit on ESC
    if (read_scancode() == 1) return 0;
  } while (1);
}

File: lens/lens.js
// script: javascript

var buf = []

var an1 = 0
var an2 = 0
const radx = 80
const rady = 30

function TIC () {
  flushBuffer(120 + Math.cos(an1)*radx,
    68 + Math.sin(an2)*rady,
    30, 30, -4, -4);
  an1 += 0.07;
  an2 += 0.13;
}

cls(0)
// for (var i = 0; i < 100; i++) {
//   line(Math.random()*240, Math.random()*136, Math.random()*240, Math.random()*136, 1);
// }
map(0, 0, 30, 17, 0, 0, 1)


for (var y = 0; y < 136; y++) {
  for (var x = 0; x < 240; x++) {
    buf.push(pix(x, y))
  }
}

function getLensBufferPixel(x, y, lensX, lensY, lensW, lensH, shiftX, shiftY) {
  var nx = x;
  var ny = y;
  var xs = x - lensX;
  var ys = y - lensY;
  var rs = lensW*lensH;
  if (xs*xs + ys*ys < rs) {
    ny = y + shiftY;
    nx = x + shiftX;
  }
  return buf[nx + ny*240]
}


function flushBuffer(lensX, lensY, lensW, lensH, shiftX, shiftY) {
  for (var y = 0; y < 136; y++) {
    for (var x = 0; x < 240; x++) {
      const color = getLensBufferPixel(x, y, lensX, lensY,
        lensW, lensH, shiftX, shiftY);
      pix(x, y, color)
    }
  }
}
